# 04.1 Origin Classification with Transformer Fine-Tuning

In [1]:
import os
import re
import json
import random
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import torch

from sklearn.metrics import (
    accuracy_score, precision_recall_fscore_support, classification_report,
    balanced_accuracy_score, confusion_matrix, ConfusionMatrixDisplay, f1_score,
)
from sklearn.model_selection import train_test_split
from sklearn.utils.class_weight import compute_class_weight
from transformers import (
    AutoModelForSequenceClassification, AutoTokenizer, DataCollatorWithPadding,
    EarlyStoppingCallback, Trainer, TrainingArguments,
)
from transformers.utils.notebook import NotebookProgressCallback

In [ ]:
RANDOM_STATE = 42
MIN_SAMPLES_PER_CLASS = 100
MODEL_CHECKPOINT = 'roberta-base'
TEXT_COLUMN = 'text_basic_plus_aspects'
MAX_LENGTH = 256

# Training config — the key changes vs. 04.1
NUM_TRAIN_EPOCHS = 12                     # was 4
TRAIN_BATCH_SIZE = 16                     # was 8
EVAL_BATCH_SIZE = 32
GRAD_ACCUM_STEPS = 2                      # effective batch = 32
WEIGHT_DECAY = 0.01
WARMUP_RATIO = 0.06                       # shorter warmup for longer runs
EARLY_STOP_PATIENCE = 3
USE_FP16 = True                           # flip to False if loss becomes NaN

# LR sweep
LR_SWEEP = [1e-5, 2e-5, 3e-5, 5e-5]

OUTPUT_DIR_ROOT = 'artifacts/origin_finetuning_roberta_fixed'
os.makedirs(OUTPUT_DIR_ROOT, exist_ok=True)

DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'
print('Device:', DEVICE)

def set_seed(seed: int):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)

set_seed(RANDOM_STATE)

## Load & preprocess

Same filtering rules, same stratified 70/15/15 split, same `text_basic_plus_aspects` column. This keeps the comparison against 04.1 and against the TF-IDF baseline clean.

In [3]:
df = pd.read_csv('Data/final_coffee_reviews.csv')
print('Shape:', df.shape)

def minimal_raw_text(text):
    text = '' if pd.isna(text) else str(text)
    return re.sub(r'\s+', ' ', text).strip().lower()

def normalize_text(text):
    text = minimal_raw_text(text)
    text = re.sub(r'[^a-z0-9\s]', ' ', text)
    return re.sub(r'\s+', ' ', text).strip()

df['text_raw_minimal'] = df['Blind Assessment'].fillna('').map(minimal_raw_text)
df['text_basic'] = df['Blind Assessment'].fillna('').map(normalize_text)

for col in ['aroma_text', 'flavor_text', 'acidity_text', 'body_text', 'combined_text']:
    if col not in df.columns:
        df[col] = ''
    df[col] = df[col].fillna('').astype(str)

df['aspect_text_combined'] = df['combined_text'].map(normalize_text)
df['text_basic_plus_aspects'] = (
    df['text_basic'].fillna('') + ' ' + df['aspect_text_combined'].fillna('')
).str.replace(r'\s+', ' ', regex=True).str.strip()

df['origin_country'] = df['Country'].astype('string').str.strip()
df['origin_country'] = df['origin_country'].replace('', pd.NA)

work = df[(df['text_raw_minimal'].str.len() >= 30) & (df['origin_country'].notna())].copy()
counts = work['origin_country'].value_counts()
valid_classes = counts[counts >= MIN_SAMPLES_PER_CLASS].index
work = work[work['origin_country'].isin(valid_classes)].copy().reset_index(drop=True)

print('Rows after filtering:', len(work))
print('Num classes:', work['origin_country'].nunique())
work['origin_country'].value_counts()

Shape: (7585, 30)
Rows after filtering: 6820
Num classes: 15


origin_country
Ethiopia         1996
Colombia          988
Kenya             699
Guatemala         561
Indonesia         452
Costa Rica        373
Panama            354
United States     302
El Salvador       240
Brazil            200
Rwanda            176
Peru              130
Nicaragua         125
Honduras          123
Mexico            101
Name: count, dtype: Int64

In [4]:
y = work['origin_country']
row_idx = work.index

train_idx, temp_idx, y_train, y_temp = train_test_split(
    row_idx, y, test_size=0.30, random_state=RANDOM_STATE, stratify=y
)
val_idx, test_idx, y_val, y_test = train_test_split(
    temp_idx, y_temp, test_size=0.50, random_state=RANDOM_STATE, stratify=y_temp
)

print('Train:', len(train_idx), 'Val:', len(val_idx), 'Test:', len(test_idx))

label_names = sorted(work['origin_country'].unique())
label2id = {label: idx for idx, label in enumerate(label_names)}
id2label = {idx: label for label, idx in label2id.items()}

train_texts = work.loc[train_idx, TEXT_COLUMN].fillna('').tolist()
val_texts   = work.loc[val_idx, TEXT_COLUMN].fillna('').tolist()
test_texts  = work.loc[test_idx, TEXT_COLUMN].fillna('').tolist()

train_labels = work.loc[train_idx, 'origin_country'].map(label2id).tolist()
val_labels   = work.loc[val_idx,   'origin_country'].map(label2id).tolist()
test_labels  = work.loc[test_idx,  'origin_country'].map(label2id).tolist()

class_weights_np = compute_class_weight(
    class_weight='balanced',
    classes=np.arange(len(label_names)),
    y=np.array(train_labels),
)
class_weights_tensor = torch.tensor(class_weights_np, dtype=torch.float)
print('Class weights (min/max):', class_weights_np.min().round(3), class_weights_np.max().round(3))

Train: 4774 Val: 1023 Test: 1023
Class weights (min/max): 0.228 4.483


In [5]:
tokenizer = AutoTokenizer.from_pretrained(MODEL_CHECKPOINT)
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

class CoffeeOriginDataset(torch.utils.data.Dataset):
    def __init__(self, texts, labels, tokenizer, max_length):
        self.encodings = tokenizer(texts, truncation=True, max_length=max_length, padding=False)
        self.labels = labels
    def __len__(self):
        return len(self.labels)
    def __getitem__(self, idx):
        item = {k: torch.tensor(v[idx]) for k, v in self.encodings.items()}
        item['labels'] = torch.tensor(self.labels[idx], dtype=torch.long)
        return item

train_dataset = CoffeeOriginDataset(train_texts, train_labels, tokenizer, MAX_LENGTH)
val_dataset   = CoffeeOriginDataset(val_texts,   val_labels,   tokenizer, MAX_LENGTH)
test_dataset  = CoffeeOriginDataset(test_texts,  test_labels,  tokenizer, MAX_LENGTH)
print('Train/Val/Test:', len(train_dataset), len(val_dataset), len(test_dataset))

Train/Val/Test: 4774 1023 1023


In [6]:
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    preds = np.argmax(logits, axis=1)
    precision, recall, f1, _ = precision_recall_fscore_support(labels, preds, average='macro', zero_division=0)
    return {
        'accuracy': accuracy_score(labels, preds),
        'balanced_accuracy': balanced_accuracy_score(labels, preds),
        'precision_macro': precision,
        'recall_macro': recall,
        'f1_macro': f1,
    }

class WeightedTrainer(Trainer):
    def __init__(self, class_weights=None, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.class_weights = class_weights
    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.pop('labels')
        outputs = model(**inputs)
        logits = outputs.get('logits')
        loss_fct = torch.nn.CrossEntropyLoss(weight=self.class_weights.to(logits.device))
        loss = loss_fct(logits, labels)
        return (loss, outputs) if return_outputs else loss

def build_training_args(output_dir: str, learning_rate: float):
    return TrainingArguments(
        output_dir=output_dir,
        learning_rate=learning_rate,
        per_device_train_batch_size=TRAIN_BATCH_SIZE,
        per_device_eval_batch_size=EVAL_BATCH_SIZE,
        gradient_accumulation_steps=GRAD_ACCUM_STEPS,
        num_train_epochs=NUM_TRAIN_EPOCHS,
        weight_decay=WEIGHT_DECAY,
        warmup_ratio=WARMUP_RATIO,
        eval_strategy='epoch',
        save_strategy='epoch',
        save_total_limit=1,
        load_best_model_at_end=True,
        metric_for_best_model='f1_macro',
        greater_is_better=True,
        logging_strategy='epoch',
        disable_tqdm=True,
        report_to='none',
        fp16=USE_FP16 and torch.cuda.is_available(),
        seed=RANDOM_STATE,
    )

def run_one(learning_rate: float, use_class_weights: bool, tag: str):
    set_seed(RANDOM_STATE)
    out_dir = os.path.join(OUTPUT_DIR_ROOT, tag)
    model = AutoModelForSequenceClassification.from_pretrained(
        MODEL_CHECKPOINT,
        num_labels=len(label_names),
        id2label=id2label,
        label2id=label2id,
    )
    args = build_training_args(out_dir, learning_rate)
    trainer_cls = WeightedTrainer if use_class_weights else Trainer
    extra = dict(class_weights=class_weights_tensor) if use_class_weights else {}
    trainer = trainer_cls(
        model=model, args=args,
        train_dataset=train_dataset, eval_dataset=val_dataset,
        processing_class=tokenizer, data_collator=data_collator,
        compute_metrics=compute_metrics,
        callbacks=[EarlyStoppingCallback(early_stopping_patience=EARLY_STOP_PATIENCE)],
        **extra,
    )
    try:
        trainer.remove_callback(NotebookProgressCallback)
    except Exception:
        pass
    print(f'\n=== Training {tag} | lr={learning_rate} | weighted={use_class_weights} ===')
    train_out = trainer.train()
    val_metrics = trainer.evaluate(eval_dataset=val_dataset)
    test_metrics = trainer.evaluate(eval_dataset=test_dataset, metric_key_prefix='test')
    return trainer, train_out, val_metrics, test_metrics

## Step 1 — LR sweep with plain (unweighted) cross-entropy

The goal here is to first verify the model can actually learn. If any LR drives training loss well below `ln(15) ≈ 2.708` and validation Macro-F1 above the TF-IDF baseline (0.134), then the 04.1 failure was a training-config issue, not a data issue.

In [7]:
sweep_results = []
for lr in LR_SWEEP:
    tag = f'plainCE_lr{lr:.0e}'
    _, train_out, val_m, test_m = run_one(lr, use_class_weights=False, tag=tag)
    sweep_results.append({
        'tag': tag,
        'lr': lr,
        'final_train_loss': train_out.training_loss,
        'val_f1_macro': val_m['eval_f1_macro'],
        'val_balanced_acc': val_m['eval_balanced_accuracy'],
        'val_accuracy': val_m['eval_accuracy'],
        'test_f1_macro': test_m['test_f1_macro'],
        'test_balanced_acc': test_m['test_balanced_accuracy'],
        'test_accuracy': test_m['test_accuracy'],
    })

sweep_df = pd.DataFrame(sweep_results).sort_values('val_f1_macro', ascending=False)
sweep_df

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training plainCE_lr1e-05 | lr=1e-05 | weighted=False ===
{'loss': '4.861', 'grad_norm': '12.66', 'learning_rate': '9.764e-06', 'epoch': '1'}
{'eval_loss': '2.3', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.406', 'eval_samples_per_second': '727.4', 'eval_steps_per_second': '22.75', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.5', 'grad_norm': '9.429', 'learning_rate': '8.877e-06', 'epoch': '2'}
{'eval_loss': '2.21', 'eval_accuracy': '0.3236', 'eval_balanced_accuracy': '0.1049', 'eval_precision_macro': '0.07844', 'eval_recall_macro': '0.1049', 'eval_f1_macro': '0.07401', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.4', 'eval_steps_per_second': '25.79', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.211', 'grad_norm': '19.37', 'learning_rate': '7.996e-06', 'epoch': '3'}
{'eval_loss': '2.136', 'eval_accuracy': '0.3372', 'eval_balanced_accuracy': '0.1415', 'eval_precision_macro': '0.1005', 'eval_recall_macro': '0.1415', 'eval_f1_macro': '0.1109', 'eval_runtime': '1.258', 'eval_samples_per_second': '813.3', 'eval_steps_per_second': '25.44', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.034', 'grad_norm': '24.89', 'learning_rate': '7.116e-06', 'epoch': '4'}
{'eval_loss': '2.14', 'eval_accuracy': '0.3646', 'eval_balanced_accuracy': '0.1415', 'eval_precision_macro': '0.1087', 'eval_recall_macro': '0.1415', 'eval_f1_macro': '0.1174', 'eval_runtime': '1.24', 'eval_samples_per_second': '825.3', 'eval_steps_per_second': '25.81', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.856', 'grad_norm': '31.63', 'learning_rate': '6.229e-06', 'epoch': '5'}
{'eval_loss': '2.143', 'eval_accuracy': '0.3431', 'eval_balanced_accuracy': '0.1454', 'eval_precision_macro': '0.09547', 'eval_recall_macro': '0.1454', 'eval_f1_macro': '0.1131', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.4', 'eval_steps_per_second': '25.79', 'epoch': '5'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.726', 'grad_norm': '42.21', 'learning_rate': '5.343e-06', 'epoch': '6'}
{'eval_loss': '2.161', 'eval_accuracy': '0.3607', 'eval_balanced_accuracy': '0.1445', 'eval_precision_macro': '0.1038', 'eval_recall_macro': '0.1445', 'eval_f1_macro': '0.1133', 'eval_runtime': '1.279', 'eval_samples_per_second': '799.7', 'eval_steps_per_second': '25.01', 'epoch': '6'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.6', 'grad_norm': '37.27', 'learning_rate': '4.456e-06', 'epoch': '7'}
{'eval_loss': '2.162', 'eval_accuracy': '0.3519', 'eval_balanced_accuracy': '0.154', 'eval_precision_macro': '0.1854', 'eval_recall_macro': '0.154', 'eval_f1_macro': '0.1335', 'eval_runtime': '1.247', 'eval_samples_per_second': '820.3', 'eval_steps_per_second': '25.66', 'epoch': '7'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.478', 'grad_norm': '36.55', 'learning_rate': '3.57e-06', 'epoch': '8'}
{'eval_loss': '2.191', 'eval_accuracy': '0.3617', 'eval_balanced_accuracy': '0.1705', 'eval_precision_macro': '0.1723', 'eval_recall_macro': '0.1705', 'eval_f1_macro': '0.143', 'eval_runtime': '1.242', 'eval_samples_per_second': '823.6', 'eval_steps_per_second': '25.76', 'epoch': '8'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.372', 'grad_norm': '41.29', 'learning_rate': '2.683e-06', 'epoch': '9'}
{'eval_loss': '2.187', 'eval_accuracy': '0.348', 'eval_balanced_accuracy': '0.16', 'eval_precision_macro': '0.1716', 'eval_recall_macro': '0.16', 'eval_f1_macro': '0.1394', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.4', 'eval_steps_per_second': '25.79', 'epoch': '9'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.279', 'grad_norm': '35.44', 'learning_rate': '1.797e-06', 'epoch': '10'}
{'eval_loss': '2.181', 'eval_accuracy': '0.3519', 'eval_balanced_accuracy': '0.1574', 'eval_precision_macro': '0.1727', 'eval_recall_macro': '0.1574', 'eval_f1_macro': '0.141', 'eval_runtime': '1.271', 'eval_samples_per_second': '804.8', 'eval_steps_per_second': '25.17', 'epoch': '10'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.218', 'grad_norm': '86.35', 'learning_rate': '9.102e-07', 'epoch': '11'}
{'eval_loss': '2.208', 'eval_accuracy': '0.3548', 'eval_balanced_accuracy': '0.1636', 'eval_precision_macro': '0.1589', 'eval_recall_macro': '0.1636', 'eval_f1_macro': '0.144', 'eval_runtime': '1.24', 'eval_samples_per_second': '824.9', 'eval_steps_per_second': '25.8', 'epoch': '11'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.161', 'grad_norm': '58.61', 'learning_rate': '2.364e-08', 'epoch': '12'}
{'eval_loss': '2.205', 'eval_accuracy': '0.3529', 'eval_balanced_accuracy': '0.1619', 'eval_precision_macro': '0.1533', 'eval_recall_macro': '0.1619', 'eval_f1_macro': '0.1445', 'eval_runtime': '1.24', 'eval_samples_per_second': '825.2', 'eval_steps_per_second': '25.81', 'epoch': '12'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'train_runtime': '373.9', 'train_samples_per_second': '153.2', 'train_steps_per_second': '4.814', 'train_loss': '3.775', 'epoch': '12'}


There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'eval_loss': '2.205', 'eval_accuracy': '0.3529', 'eval_balanced_accuracy': '0.1619', 'eval_precision_macro': '0.1533', 'eval_recall_macro': '0.1619', 'eval_f1_macro': '0.1445', 'eval_runtime': '1.461', 'eval_samples_per_second': '700.2', 'eval_steps_per_second': '21.9', 'epoch': '12'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.136', 'test_accuracy': '0.3773', 'test_balanced_accuracy': '0.1708', 'test_precision_macro': '0.1701', 'test_recall_macro': '0.1708', 'test_f1_macro': '0.1516', 'test_runtime': '1.309', 'test_samples_per_second': '781.5', 'test_steps_per_second': '24.45', 'epoch': '12'}


Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training plainCE_lr2e-05 | lr=2e-05 | weighted=False ===
{'loss': '4.797', 'grad_norm': 'inf', 'learning_rate': '1.953e-05', 'epoch': '1'}
{'eval_loss': '2.312', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.238', 'eval_samples_per_second': '826.1', 'eval_steps_per_second': '25.84', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.502', 'grad_norm': '11.69', 'learning_rate': '1.777e-05', 'epoch': '2'}
{'eval_loss': '2.247', 'eval_accuracy': '0.2962', 'eval_balanced_accuracy': '0.09739', 'eval_precision_macro': '0.05974', 'eval_recall_macro': '0.09739', 'eval_f1_macro': '0.05583', 'eval_runtime': '1.249', 'eval_samples_per_second': '819', 'eval_steps_per_second': '25.62', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.19', 'grad_norm': '13', 'learning_rate': '1.599e-05', 'epoch': '3'}
{'eval_loss': '2.138', 'eval_accuracy': '0.3421', 'eval_balanced_accuracy': '0.1429', 'eval_precision_macro': '0.1339', 'eval_recall_macro': '0.1429', 'eval_f1_macro': '0.1113', 'eval_runtime': '1.244', 'eval_samples_per_second': '822.2', 'eval_steps_per_second': '25.72', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.996', 'grad_norm': '19.04', 'learning_rate': '1.422e-05', 'epoch': '4'}
{'eval_loss': '2.132', 'eval_accuracy': '0.3656', 'eval_balanced_accuracy': '0.1421', 'eval_precision_macro': '0.1181', 'eval_recall_macro': '0.1421', 'eval_f1_macro': '0.119', 'eval_runtime': '1.246', 'eval_samples_per_second': '821.1', 'eval_steps_per_second': '25.68', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.787', 'grad_norm': '19.09', 'learning_rate': '1.245e-05', 'epoch': '5'}
{'eval_loss': '2.152', 'eval_accuracy': '0.3509', 'eval_balanced_accuracy': '0.1526', 'eval_precision_macro': '0.1249', 'eval_recall_macro': '0.1526', 'eval_f1_macro': '0.1225', 'eval_runtime': '1.273', 'eval_samples_per_second': '803.6', 'eval_steps_per_second': '25.14', 'epoch': '5'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.594', 'grad_norm': '26.07', 'learning_rate': '1.067e-05', 'epoch': '6'}
{'eval_loss': '2.169', 'eval_accuracy': '0.3705', 'eval_balanced_accuracy': '0.1528', 'eval_precision_macro': '0.1449', 'eval_recall_macro': '0.1528', 'eval_f1_macro': '0.1255', 'eval_runtime': '1.252', 'eval_samples_per_second': '817.3', 'eval_steps_per_second': '25.57', 'epoch': '6'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.391', 'grad_norm': '32.35', 'learning_rate': '8.901e-06', 'epoch': '7'}
{'eval_loss': '2.216', 'eval_accuracy': '0.3548', 'eval_balanced_accuracy': '0.1617', 'eval_precision_macro': '0.1482', 'eval_recall_macro': '0.1617', 'eval_f1_macro': '0.1423', 'eval_runtime': '1.254', 'eval_samples_per_second': '816.1', 'eval_steps_per_second': '25.53', 'epoch': '7'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.206', 'grad_norm': '32.56', 'learning_rate': '7.139e-06', 'epoch': '8'}
{'eval_loss': '2.216', 'eval_accuracy': '0.3636', 'eval_balanced_accuracy': '0.1653', 'eval_precision_macro': '0.156', 'eval_recall_macro': '0.1653', 'eval_f1_macro': '0.1479', 'eval_runtime': '1.239', 'eval_samples_per_second': '825.9', 'eval_steps_per_second': '25.84', 'epoch': '8'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.031', 'grad_norm': '45.26', 'learning_rate': '5.366e-06', 'epoch': '9'}
{'eval_loss': '2.265', 'eval_accuracy': '0.3685', 'eval_balanced_accuracy': '0.1756', 'eval_precision_macro': '0.1696', 'eval_recall_macro': '0.1756', 'eval_f1_macro': '0.1589', 'eval_runtime': '1.239', 'eval_samples_per_second': '825.5', 'eval_steps_per_second': '25.82', 'epoch': '9'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.871', 'grad_norm': '31.75', 'learning_rate': '3.593e-06', 'epoch': '10'}
{'eval_loss': '2.302', 'eval_accuracy': '0.3597', 'eval_balanced_accuracy': '0.1745', 'eval_precision_macro': '0.1727', 'eval_recall_macro': '0.1745', 'eval_f1_macro': '0.1644', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.6', 'eval_steps_per_second': '25.79', 'epoch': '10'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.756', 'grad_norm': '53.36', 'learning_rate': '1.82e-06', 'epoch': '11'}
{'eval_loss': '2.345', 'eval_accuracy': '0.3587', 'eval_balanced_accuracy': '0.1708', 'eval_precision_macro': '0.1748', 'eval_recall_macro': '0.1708', 'eval_f1_macro': '0.1591', 'eval_runtime': '1.238', 'eval_samples_per_second': '826', 'eval_steps_per_second': '25.84', 'epoch': '11'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.675', 'grad_norm': '49.03', 'learning_rate': '4.728e-08', 'epoch': '12'}
{'eval_loss': '2.345', 'eval_accuracy': '0.3578', 'eval_balanced_accuracy': '0.174', 'eval_precision_macro': '0.1829', 'eval_recall_macro': '0.174', 'eval_f1_macro': '0.1643', 'eval_runtime': '1.238', 'eval_samples_per_second': '826.4', 'eval_steps_per_second': '25.85', 'epoch': '12'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'train_runtime': '359.1', 'train_samples_per_second': '159.5', 'train_steps_per_second': '5.012', 'train_loss': '3.567', 'epoch': '12'}


There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'eval_loss': '2.302', 'eval_accuracy': '0.3597', 'eval_balanced_accuracy': '0.1745', 'eval_precision_macro': '0.1726', 'eval_recall_macro': '0.1745', 'eval_f1_macro': '0.1644', 'eval_runtime': '1.585', 'eval_samples_per_second': '645.3', 'eval_steps_per_second': '20.19', 'epoch': '12'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.187', 'test_accuracy': '0.3646', 'test_balanced_accuracy': '0.1712', 'test_precision_macro': '0.1594', 'test_recall_macro': '0.1712', 'test_f1_macro': '0.1583', 'test_runtime': '1.306', 'test_samples_per_second': '783.5', 'test_steps_per_second': '24.51', 'epoch': '12'}


Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training plainCE_lr3e-05 | lr=3e-05 | weighted=False ===
{'loss': '4.762', 'grad_norm': '14.26', 'learning_rate': '2.933e-05', 'epoch': '1'}
{'eval_loss': '2.287', 'eval_accuracy': '0.2815', 'eval_balanced_accuracy': '0.06686', 'eval_precision_macro': '0.02372', 'eval_recall_macro': '0.06686', 'eval_f1_macro': '0.03451', 'eval_runtime': '1.244', 'eval_samples_per_second': '822.1', 'eval_steps_per_second': '25.72', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.412', 'grad_norm': '7.017', 'learning_rate': '2.667e-05', 'epoch': '2'}
{'eval_loss': '2.179', 'eval_accuracy': '0.3382', 'eval_balanced_accuracy': '0.1026', 'eval_precision_macro': '0.06591', 'eval_recall_macro': '0.1026', 'eval_f1_macro': '0.07055', 'eval_runtime': '1.248', 'eval_samples_per_second': '819.9', 'eval_steps_per_second': '25.65', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.096', 'grad_norm': '10.87', 'learning_rate': '2.401e-05', 'epoch': '3'}
{'eval_loss': '2.109', 'eval_accuracy': '0.3304', 'eval_balanced_accuracy': '0.1498', 'eval_precision_macro': '0.1145', 'eval_recall_macro': '0.1498', 'eval_f1_macro': '0.1251', 'eval_runtime': '1.323', 'eval_samples_per_second': '773', 'eval_steps_per_second': '24.18', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.891', 'grad_norm': '17.95', 'learning_rate': '2.135e-05', 'epoch': '4'}
{'eval_loss': '2.141', 'eval_accuracy': '0.3627', 'eval_balanced_accuracy': '0.1456', 'eval_precision_macro': '0.1047', 'eval_recall_macro': '0.1456', 'eval_f1_macro': '0.1188', 'eval_runtime': '1.257', 'eval_samples_per_second': '814', 'eval_steps_per_second': '25.46', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.649', 'grad_norm': '24.12', 'learning_rate': '1.869e-05', 'epoch': '5'}
{'eval_loss': '2.156', 'eval_accuracy': '0.3587', 'eval_balanced_accuracy': '0.1616', 'eval_precision_macro': '0.1539', 'eval_recall_macro': '0.1616', 'eval_f1_macro': '0.1412', 'eval_runtime': '1.249', 'eval_samples_per_second': '819.2', 'eval_steps_per_second': '25.63', 'epoch': '5'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.367', 'grad_norm': '27.63', 'learning_rate': '1.603e-05', 'epoch': '6'}
{'eval_loss': '2.208', 'eval_accuracy': '0.3568', 'eval_balanced_accuracy': '0.1619', 'eval_precision_macro': '0.1729', 'eval_recall_macro': '0.1619', 'eval_f1_macro': '0.1438', 'eval_runtime': '1.238', 'eval_samples_per_second': '826', 'eval_steps_per_second': '25.84', 'epoch': '6'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.114', 'grad_norm': '29.04', 'learning_rate': '1.337e-05', 'epoch': '7'}
{'eval_loss': '2.223', 'eval_accuracy': '0.3451', 'eval_balanced_accuracy': '0.167', 'eval_precision_macro': '0.1407', 'eval_recall_macro': '0.167', 'eval_f1_macro': '0.1499', 'eval_runtime': '1.239', 'eval_samples_per_second': '825.4', 'eval_steps_per_second': '25.82', 'epoch': '7'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.834', 'grad_norm': '35.1', 'learning_rate': '1.071e-05', 'epoch': '8'}
{'eval_loss': '2.29', 'eval_accuracy': '0.346', 'eval_balanced_accuracy': '0.1804', 'eval_precision_macro': '0.158', 'eval_recall_macro': '0.1804', 'eval_f1_macro': '0.1618', 'eval_runtime': '1.239', 'eval_samples_per_second': '825.7', 'eval_steps_per_second': '25.83', 'epoch': '8'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.595', 'grad_norm': '45.98', 'learning_rate': '8.05e-06', 'epoch': '9'}
{'eval_loss': '2.395', 'eval_accuracy': '0.3412', 'eval_balanced_accuracy': '0.1745', 'eval_precision_macro': '0.1609', 'eval_recall_macro': '0.1745', 'eval_f1_macro': '0.1591', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.4', 'eval_steps_per_second': '25.79', 'epoch': '9'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.363', 'grad_norm': '28.92', 'learning_rate': '5.39e-06', 'epoch': '10'}
{'eval_loss': '2.406', 'eval_accuracy': '0.35', 'eval_balanced_accuracy': '0.1812', 'eval_precision_macro': '0.1934', 'eval_recall_macro': '0.1812', 'eval_f1_macro': '0.1734', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.1', 'eval_steps_per_second': '25.78', 'epoch': '10'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.193', 'grad_norm': '57.92', 'learning_rate': '2.73e-06', 'epoch': '11'}
{'eval_loss': '2.496', 'eval_accuracy': '0.3539', 'eval_balanced_accuracy': '0.1769', 'eval_precision_macro': '0.1684', 'eval_recall_macro': '0.1769', 'eval_f1_macro': '0.1662', 'eval_runtime': '1.242', 'eval_samples_per_second': '823.9', 'eval_steps_per_second': '25.77', 'epoch': '11'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.045', 'grad_norm': '28.78', 'learning_rate': '7.092e-08', 'epoch': '12'}
{'eval_loss': '2.512', 'eval_accuracy': '0.3509', 'eval_balanced_accuracy': '0.1876', 'eval_precision_macro': '0.2068', 'eval_recall_macro': '0.1876', 'eval_f1_macro': '0.1815', 'eval_runtime': '1.244', 'eval_samples_per_second': '822.2', 'eval_steps_per_second': '25.72', 'epoch': '12'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'train_runtime': '366.5', 'train_samples_per_second': '156.3', 'train_steps_per_second': '4.911', 'train_loss': '3.277', 'epoch': '12'}


There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'eval_loss': '2.512', 'eval_accuracy': '0.3509', 'eval_balanced_accuracy': '0.1876', 'eval_precision_macro': '0.2068', 'eval_recall_macro': '0.1876', 'eval_f1_macro': '0.1815', 'eval_runtime': '1.523', 'eval_samples_per_second': '671.9', 'eval_steps_per_second': '21.02', 'epoch': '12'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.431', 'test_accuracy': '0.3529', 'test_balanced_accuracy': '0.1748', 'test_precision_macro': '0.1588', 'test_recall_macro': '0.1748', 'test_f1_macro': '0.1644', 'test_runtime': '1.348', 'test_samples_per_second': '759', 'test_steps_per_second': '23.74', 'epoch': '12'}


Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training plainCE_lr5e-05 | lr=5e-05 | weighted=False ===
{'loss': '4.751', 'grad_norm': '8.646', 'learning_rate': '4.888e-05', 'epoch': '1'}
{'eval_loss': '2.332', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.249', 'eval_samples_per_second': '818.8', 'eval_steps_per_second': '25.61', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.647', 'grad_norm': '4.925', 'learning_rate': '4.444e-05', 'epoch': '2'}
{'eval_loss': '2.318', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.247', 'eval_samples_per_second': '820.4', 'eval_steps_per_second': '25.66', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.644', 'grad_norm': '6.085', 'learning_rate': '4.001e-05', 'epoch': '3'}
{'eval_loss': '2.314', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.265', 'eval_samples_per_second': '808.5', 'eval_steps_per_second': '25.29', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.634', 'grad_norm': '5.108', 'learning_rate': '3.558e-05', 'epoch': '4'}
{'eval_loss': '2.317', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.237', 'eval_samples_per_second': '826.9', 'eval_steps_per_second': '25.87', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'train_runtime': '118.3', 'train_samples_per_second': '484.4', 'train_steps_per_second': '15.22', 'train_loss': '4.669', 'epoch': '4'}


There were unexpected keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.beta', 'roberta.embeddings.LayerNorm.gamma', 'roberta.encoder.layer.0.attention.output.LayerNorm.beta', 'roberta.encoder.layer.0.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.0.output.LayerNorm.beta', 'roberta.encoder.layer.0.output.LayerNorm.gamma', 'roberta.encoder.layer.1.attention.output.LayerNorm.beta', 'roberta.encoder.layer.1.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.1.output.LayerNorm.beta', 'roberta.encoder.layer.1.output.LayerNorm.gamma', 'roberta.encoder.layer.2.attention.output.LayerNorm.beta', 'roberta.encoder.layer.2.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.2.output.LayerNorm.beta', 'roberta.encoder.layer.2.output.LayerNorm.gamma', 'roberta.encoder.layer.3.attention.output.LayerNorm.beta', 'roberta.encoder.layer.3.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.3.output.LayerNorm.beta', 'roberta.encoder.layer.3.output.LayerNorm.g

{'eval_loss': '2.331', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.493', 'eval_samples_per_second': '685.1', 'eval_steps_per_second': '21.43', 'epoch': '4'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.333', 'test_accuracy': '0.2923', 'test_balanced_accuracy': '0.06667', 'test_precision_macro': '0.01949', 'test_recall_macro': '0.06667', 'test_f1_macro': '0.03016', 'test_runtime': '1.317', 'test_samples_per_second': '776.4', 'test_steps_per_second': '24.29', 'epoch': '4'}


,tag,lr,final_train_loss,val_f1_macro,val_balanced_acc,val_accuracy,test_f1_macro,test_balanced_acc,test_accuracy
2,plainCE_lr3e-05,0.00003,3.276758,0.181486,0.187649,0.350929,0.164361,0.174847,0.352884
1,plainCE_lr2e-05,0.00002,3.566520,0.164377,0.174513,0.359726,0.158272,0.171164,0.364614
0,plainCE_lr1e-05,0.00001,3.774636,0.144540,0.161936,0.352884,0.151584,0.170846,0.377322
3,plainCE_lr5e-05,0.00005,4.669088,0.030234,0.066667,0.293255,0.030156,0.066667,0.292278


In [8]:
best_row = sweep_df.iloc[0]
BEST_LR = float(best_row['lr'])
print(f'Best LR (by val macro-F1): {BEST_LR}')
print(f'  val macro-F1:  {best_row["val_f1_macro"]:.4f}')
print(f'  test macro-F1: {best_row["test_f1_macro"]:.4f}')
print(f'  test bal-acc:  {best_row["test_balanced_acc"]:.4f}')
print(f'  TF-IDF baseline (from notebook 02): 0.134 macro-F1')

Best LR (by val macro-F1): 3e-05
  val macro-F1:  0.1815
  test macro-F1: 0.1644
  test bal-acc:  0.1748
  TF-IDF baseline (from notebook 02): 0.134 macro-F1


## Step 2 — Rerun best LR with class-weighted CE

Now that we know the model can learn, we re-introduce class weighting to see if it improves macro-F1 on tail classes without collapsing performance on head classes.

In [9]:
weighted_trainer, weighted_train_out, weighted_val_m, weighted_test_m = run_one(
    BEST_LR, use_class_weights=True, tag=f'weightedCE_lr{BEST_LR:.0e}'
)

final_summary = pd.DataFrame([
    {
        'setup': f'plain CE @ lr={BEST_LR}',
        'val_f1_macro': best_row['val_f1_macro'],
        'val_bal_acc': best_row['val_balanced_acc'],
        'test_f1_macro': best_row['test_f1_macro'],
        'test_bal_acc': best_row['test_balanced_acc'],
        'test_accuracy': best_row['test_accuracy'],
    },
    {
        'setup': f'weighted CE @ lr={BEST_LR}',
        'val_f1_macro': weighted_val_m['eval_f1_macro'],
        'val_bal_acc': weighted_val_m['eval_balanced_accuracy'],
        'test_f1_macro': weighted_test_m['test_f1_macro'],
        'test_bal_acc': weighted_test_m['test_balanced_accuracy'],
        'test_accuracy': weighted_test_m['test_accuracy'],
    },
])
final_summary

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training weightedCE_lr3e-05 | lr=3e-05 | weighted=True ===
{'loss': '5.411', 'grad_norm': '5.057', 'learning_rate': '2.927e-05', 'epoch': '1'}
{'eval_loss': '2.709', 'eval_accuracy': '0.06745', 'eval_balanced_accuracy': '0.07804', 'eval_precision_macro': '0.009934', 'eval_recall_macro': '0.07804', 'eval_f1_macro': '0.01708', 'eval_runtime': '1.249', 'eval_samples_per_second': '819', 'eval_steps_per_second': '25.62', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.41', 'grad_norm': '3.837', 'learning_rate': '2.661e-05', 'epoch': '2'}
{'eval_loss': '2.709', 'eval_accuracy': '0.1026', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.006843', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.01241', 'eval_runtime': '1.277', 'eval_samples_per_second': '801.2', 'eval_steps_per_second': '25.06', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.408', 'grad_norm': '5.277', 'learning_rate': '2.395e-05', 'epoch': '3'}
{'eval_loss': '2.711', 'eval_accuracy': '0.1026', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.006843', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.01241', 'eval_runtime': '1.256', 'eval_samples_per_second': '814.7', 'eval_steps_per_second': '25.48', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.402', 'grad_norm': '5.292', 'learning_rate': '2.129e-05', 'epoch': '4'}
{'eval_loss': '2.704', 'eval_accuracy': '0.2522', 'eval_balanced_accuracy': '0.09676', 'eval_precision_macro': '0.03094', 'eval_recall_macro': '0.09676', 'eval_f1_macro': '0.0446', 'eval_runtime': '1.243', 'eval_samples_per_second': '822.9', 'eval_steps_per_second': '25.74', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.407', 'grad_norm': '2.714', 'learning_rate': '1.863e-05', 'epoch': '5'}
{'eval_loss': '2.709', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.281', 'eval_samples_per_second': '798.3', 'eval_steps_per_second': '24.97', 'epoch': '5'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.4', 'grad_norm': '7.442', 'learning_rate': '1.598e-05', 'epoch': '6'}
{'eval_loss': '2.709', 'eval_accuracy': '0.1026', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.006843', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.01241', 'eval_runtime': '1.292', 'eval_samples_per_second': '792.1', 'eval_steps_per_second': '24.78', 'epoch': '6'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.404', 'grad_norm': '4.217', 'learning_rate': '1.332e-05', 'epoch': '7'}
{'eval_loss': '2.708', 'eval_accuracy': '0.2933', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.01955', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.03023', 'eval_runtime': '1.295', 'eval_samples_per_second': '790.1', 'eval_steps_per_second': '24.72', 'epoch': '7'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'train_runtime': '211.9', 'train_samples_per_second': '270.3', 'train_steps_per_second': '8.493', 'train_loss': '5.406', 'epoch': '7'}


There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'eval_loss': '2.704', 'eval_accuracy': '0.2502', 'eval_balanced_accuracy': '0.09783', 'eval_precision_macro': '0.03108', 'eval_recall_macro': '0.09783', 'eval_f1_macro': '0.04462', 'eval_runtime': '1.261', 'eval_samples_per_second': '811', 'eval_steps_per_second': '25.37', 'epoch': '7'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.699', 'test_accuracy': '0.2307', 'test_balanced_accuracy': '0.08292', 'test_precision_macro': '0.02845', 'test_recall_macro': '0.08292', 'test_f1_macro': '0.04012', 'test_runtime': '1.355', 'test_samples_per_second': '755.3', 'test_steps_per_second': '23.63', 'epoch': '7'}


,setup,val_f1_macro,val_bal_acc,test_f1_macro,test_bal_acc,test_accuracy
0,plain CE @ lr=3e-05,0.181486,0.187649,0.164361,0.174847,0.352884
1,weighted CE @ lr=3e-05,0.044618,0.097830,0.040115,0.082917,0.230694


## Step 3 — Diagnostics on the best model

Pick whichever setup won on val macro-F1 and produce a per-class report plus confusion matrix.

In [10]:
# Choose winner by val macro-F1
if weighted_val_m['eval_f1_macro'] >= best_row['val_f1_macro']:
    winning_trainer = weighted_trainer
    winning_name = f'weighted CE @ lr={BEST_LR}'
else:
    # Re-train best plain-CE to get its trainer object in scope
    winning_trainer, _, _, _ = run_one(BEST_LR, use_class_weights=False, tag=f'plainCE_best_rerun_lr{BEST_LR:.0e}')
    winning_name = f'plain CE @ lr={BEST_LR}'

print(f'Winning setup: {winning_name}')

val_pred_output  = winning_trainer.predict(val_dataset)
test_pred_output = winning_trainer.predict(test_dataset)
val_pred_ids  = np.argmax(val_pred_output.predictions, axis=1)
test_pred_ids = np.argmax(test_pred_output.predictions, axis=1)

val_true_labels  = [id2label[i] for i in val_labels]
test_true_labels = [id2label[i] for i in test_labels]
val_pred_labels  = [id2label[i] for i in val_pred_ids]
test_pred_labels = [id2label[i] for i in test_pred_ids]

print('\n== Test headline ==')
print(f'Macro-F1:          {f1_score(test_true_labels, test_pred_labels, average="macro", zero_division=0):.4f}')
print(f'Balanced Accuracy: {balanced_accuracy_score(test_true_labels, test_pred_labels):.4f}')
print(f'Accuracy:          {accuracy_score(test_true_labels, test_pred_labels):.4f}  (biased by imbalance)')

class_order = work['origin_country'].value_counts().index.tolist()
print('\nTest per-class report (sorted by support):')
print(classification_report(test_true_labels, test_pred_labels, labels=class_order, zero_division=0, digits=3))

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training plainCE_best_rerun_lr3e-05 | lr=3e-05 | weighted=False ===
{'loss': '4.745', 'grad_norm': '15.88', 'learning_rate': '2.929e-05', 'epoch': '1'}
{'eval_loss': '2.279', 'eval_accuracy': '0.2717', 'eval_balanced_accuracy': '0.06978', 'eval_precision_macro': '0.02681', 'eval_recall_macro': '0.06978', 'eval_f1_macro': '0.03873', 'eval_runtime': '1.277', 'eval_samples_per_second': '800.8', 'eval_steps_per_second': '25.05', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.334', 'grad_norm': '7.633', 'learning_rate': '2.663e-05', 'epoch': '2'}
{'eval_loss': '2.158', 'eval_accuracy': '0.3431', 'eval_balanced_accuracy': '0.1184', 'eval_precision_macro': '0.1392', 'eval_recall_macro': '0.1184', 'eval_f1_macro': '0.09953', 'eval_runtime': '1.243', 'eval_samples_per_second': '823.1', 'eval_steps_per_second': '25.75', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.076', 'grad_norm': '10.64', 'learning_rate': '2.397e-05', 'epoch': '3'}
{'eval_loss': '2.095', 'eval_accuracy': '0.3412', 'eval_balanced_accuracy': '0.1486', 'eval_precision_macro': '0.1252', 'eval_recall_macro': '0.1486', 'eval_f1_macro': '0.1285', 'eval_runtime': '1.243', 'eval_samples_per_second': '822.8', 'eval_steps_per_second': '25.74', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.883', 'grad_norm': '15.04', 'learning_rate': '2.131e-05', 'epoch': '4'}
{'eval_loss': '2.103', 'eval_accuracy': '0.3675', 'eval_balanced_accuracy': '0.1516', 'eval_precision_macro': '0.1148', 'eval_recall_macro': '0.1516', 'eval_f1_macro': '0.1281', 'eval_runtime': '1.264', 'eval_samples_per_second': '809.6', 'eval_steps_per_second': '25.32', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.627', 'grad_norm': '31.11', 'learning_rate': '1.865e-05', 'epoch': '5'}
{'eval_loss': '2.17', 'eval_accuracy': '0.3519', 'eval_balanced_accuracy': '0.1671', 'eval_precision_macro': '0.1695', 'eval_recall_macro': '0.1671', 'eval_f1_macro': '0.1427', 'eval_runtime': '1.247', 'eval_samples_per_second': '820.4', 'eval_steps_per_second': '25.66', 'epoch': '5'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.356', 'grad_norm': '30.35', 'learning_rate': '1.601e-05', 'epoch': '6'}
{'eval_loss': '2.178', 'eval_accuracy': '0.3656', 'eval_balanced_accuracy': '0.1647', 'eval_precision_macro': '0.1558', 'eval_recall_macro': '0.1647', 'eval_f1_macro': '0.142', 'eval_runtime': '1.255', 'eval_samples_per_second': '815.4', 'eval_steps_per_second': '25.51', 'epoch': '6'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.091', 'grad_norm': '33.82', 'learning_rate': '1.335e-05', 'epoch': '7'}
{'eval_loss': '2.238', 'eval_accuracy': '0.3675', 'eval_balanced_accuracy': '0.1578', 'eval_precision_macro': '0.1804', 'eval_recall_macro': '0.1578', 'eval_f1_macro': '0.1499', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.2', 'eval_steps_per_second': '25.78', 'epoch': '7'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.816', 'grad_norm': '53.9', 'learning_rate': '1.069e-05', 'epoch': '8'}
{'eval_loss': '2.298', 'eval_accuracy': '0.348', 'eval_balanced_accuracy': '0.173', 'eval_precision_macro': '0.1814', 'eval_recall_macro': '0.173', 'eval_f1_macro': '0.1569', 'eval_runtime': '1.287', 'eval_samples_per_second': '794.7', 'eval_steps_per_second': '24.86', 'epoch': '8'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.581', 'grad_norm': '48.31', 'learning_rate': '8.032e-06', 'epoch': '9'}
{'eval_loss': '2.353', 'eval_accuracy': '0.3666', 'eval_balanced_accuracy': '0.1818', 'eval_precision_macro': '0.2532', 'eval_recall_macro': '0.1818', 'eval_f1_macro': '0.1794', 'eval_runtime': '1.251', 'eval_samples_per_second': '817.9', 'eval_steps_per_second': '25.58', 'epoch': '9'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.352', 'grad_norm': '28.86', 'learning_rate': '5.372e-06', 'epoch': '10'}
{'eval_loss': '2.407', 'eval_accuracy': '0.3558', 'eval_balanced_accuracy': '0.1812', 'eval_precision_macro': '0.2939', 'eval_recall_macro': '0.1812', 'eval_f1_macro': '0.1805', 'eval_runtime': '1.241', 'eval_samples_per_second': '824.2', 'eval_steps_per_second': '25.78', 'epoch': '10'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.181', 'grad_norm': 'inf', 'learning_rate': '2.713e-06', 'epoch': '11'}
{'eval_loss': '2.444', 'eval_accuracy': '0.3666', 'eval_balanced_accuracy': '0.1858', 'eval_precision_macro': '0.2062', 'eval_recall_macro': '0.1858', 'eval_f1_macro': '0.1801', 'eval_runtime': '1.256', 'eval_samples_per_second': '814.6', 'eval_steps_per_second': '25.48', 'epoch': '11'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.034', 'grad_norm': '39.83', 'learning_rate': '7.092e-08', 'epoch': '12'}
{'eval_loss': '2.467', 'eval_accuracy': '0.3539', 'eval_balanced_accuracy': '0.1813', 'eval_precision_macro': '0.1876', 'eval_recall_macro': '0.1813', 'eval_f1_macro': '0.174', 'eval_runtime': '1.251', 'eval_samples_per_second': '817.9', 'eval_steps_per_second': '25.58', 'epoch': '12'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'train_runtime': '356.5', 'train_samples_per_second': '160.7', 'train_steps_per_second': '5.049', 'train_loss': '3.256', 'epoch': '12'}


There were unexpected keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.beta', 'roberta.embeddings.LayerNorm.gamma', 'roberta.encoder.layer.0.attention.output.LayerNorm.beta', 'roberta.encoder.layer.0.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.0.output.LayerNorm.beta', 'roberta.encoder.layer.0.output.LayerNorm.gamma', 'roberta.encoder.layer.1.attention.output.LayerNorm.beta', 'roberta.encoder.layer.1.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.1.output.LayerNorm.beta', 'roberta.encoder.layer.1.output.LayerNorm.gamma', 'roberta.encoder.layer.2.attention.output.LayerNorm.beta', 'roberta.encoder.layer.2.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.2.output.LayerNorm.beta', 'roberta.encoder.layer.2.output.LayerNorm.gamma', 'roberta.encoder.layer.3.attention.output.LayerNorm.beta', 'roberta.encoder.layer.3.attention.output.LayerNorm.gamma', 'roberta.encoder.layer.3.output.LayerNorm.beta', 'roberta.encoder.layer.3.output.LayerNorm.g

{'eval_loss': '2.408', 'eval_accuracy': '0.3539', 'eval_balanced_accuracy': '0.1795', 'eval_precision_macro': '0.2792', 'eval_recall_macro': '0.1795', 'eval_f1_macro': '0.1771', 'eval_runtime': '1.564', 'eval_samples_per_second': '654.1', 'eval_steps_per_second': '20.46', 'epoch': '12'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.35', 'test_accuracy': '0.3675', 'test_balanced_accuracy': '0.1849', 'test_precision_macro': '0.225', 'test_recall_macro': '0.1849', 'test_f1_macro': '0.1783', 'test_runtime': '1.334', 'test_samples_per_second': '766.9', 'test_steps_per_second': '23.99', 'epoch': '12'}
Winning setup: plain CE @ lr=3e-05

== Test headline ==
Macro-F1:          0.1783
Balanced Accuracy: 0.1849
Accuracy:          0.3675  (biased by imbalance)

Test per-class report (sorted by support):
               precision    recall  f1-score   support

     Ethiopia      0.454     0.632     0.529       299
     Colombia      0.254     0.318     0.282       148
        Kenya      0.491     0.533     0.511       105
    Guatemala      0.231     0.405     0.294        84
    Indonesia      0.529     0.529     0.529        68
   Costa Rica      0.088     0.054     0.067        56
       Panama      0.160     0.075     0.103        53
United States      0.167     0.022     0.039        45
  El Salvador    

In [ ]:
cm = confusion_matrix(test_true_labels, test_pred_labels, labels=class_order)
fig, ax = plt.subplots(figsize=(10, 8))
ConfusionMatrixDisplay(confusion_matrix=cm, display_labels=class_order).plot(
    ax=ax, xticks_rotation=90, colorbar=False,
)
plt.title(f'Test Confusion Matrix — {winning_name}')
plt.tight_layout()
plt.show()

In [12]:
results_path = os.path.join(OUTPUT_DIR_ROOT, 'sweep_results.json')
with open(results_path, 'w') as f:
    json.dump({
        'sweep': sweep_results,
        'weighted_best': {
            'lr': BEST_LR,
            'val_f1_macro': weighted_val_m['eval_f1_macro'],
            'test_f1_macro': weighted_test_m['test_f1_macro'],
            'test_balanced_accuracy': weighted_test_m['test_balanced_accuracy'],
        },
        'winner': winning_name,
        'tfidf_baseline_f1_macro': 0.134,
    }, f, indent=2)
print('Saved:', results_path)

Saved: artifacts/origin_finetuning_roberta_fixed\sweep_results.json


## How to interpret the output

- **Training loss should drop clearly below 2.70.** If it stays near 2.70 across all LRs, the problem is not LR — it's the data or the model head.
- **At least one LR should beat the TF-IDF baseline (test Macro-F1 > 0.134).** If not, that is a negative result worth reporting honestly: transformers did not help on this dataset at this scale.
- **Weighted CE usually lifts macro-F1 a little and hurts accuracy.** If weighted CE collapses (macro-F1 ≈ 0.06 again), the class weights are destabilizing training — halve them or switch strategies (e.g., focal loss).

## Step 4 — Weighted CE at lower LRs (fix for collapse)

Weighted CE at lr=5e-5 collapsed (Macro-F1 ≈ 0.012, balanced acc = 1/15 → predicting a single class). High LR × strong class weights destabilized training. Retry at lr=1e-5 and 2e-5.

In [13]:
weighted_retry_results = []
for lr in [1e-5, 2e-5]:
    tag = f'weightedCE_retry_lr{lr:.0e}'
    _, train_out, val_m, test_m = run_one(lr, use_class_weights=True, tag=tag)
    weighted_retry_results.append({
        'tag': tag,
        'lr': lr,
        'final_train_loss': train_out.training_loss,
        'val_f1_macro': val_m['eval_f1_macro'],
        'val_balanced_acc': val_m['eval_balanced_accuracy'],
        'val_accuracy': val_m['eval_accuracy'],
        'test_f1_macro': test_m['test_f1_macro'],
        'test_balanced_acc': test_m['test_balanced_accuracy'],
        'test_accuracy': test_m['test_accuracy'],
    })

weighted_retry_df = pd.DataFrame(weighted_retry_results).sort_values('val_f1_macro', ascending=False)
weighted_retry_df

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training weightedCE_retry_lr1e-05 | lr=1e-05 | weighted=True ===
{'loss': '5.409', 'grad_norm': '5.459', 'learning_rate': '9.758e-06', 'epoch': '1'}
{'eval_loss': '2.708', 'eval_accuracy': '0.08211', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.005474', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.01012', 'eval_runtime': '1.278', 'eval_samples_per_second': '800.6', 'eval_steps_per_second': '25.04', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.378', 'grad_norm': '9.941', 'learning_rate': '8.871e-06', 'epoch': '2'}
{'eval_loss': '2.658', 'eval_accuracy': '0.1486', 'eval_balanced_accuracy': '0.1173', 'eval_precision_macro': '0.1304', 'eval_recall_macro': '0.1173', 'eval_f1_macro': '0.06317', 'eval_runtime': '1.246', 'eval_samples_per_second': '821.1', 'eval_steps_per_second': '25.68', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.137', 'grad_norm': '24.72', 'learning_rate': '7.996e-06', 'epoch': '3'}
{'eval_loss': '2.555', 'eval_accuracy': '0.1691', 'eval_balanced_accuracy': '0.1721', 'eval_precision_macro': '0.2163', 'eval_recall_macro': '0.1721', 'eval_f1_macro': '0.1109', 'eval_runtime': '1.258', 'eval_samples_per_second': '813.5', 'eval_steps_per_second': '25.45', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.867', 'grad_norm': '44.45', 'learning_rate': '7.11e-06', 'epoch': '4'}
{'eval_loss': '2.5', 'eval_accuracy': '0.2131', 'eval_balanced_accuracy': '0.1918', 'eval_precision_macro': '0.1574', 'eval_recall_macro': '0.1918', 'eval_f1_macro': '0.1507', 'eval_runtime': '1.281', 'eval_samples_per_second': '798.4', 'eval_steps_per_second': '24.97', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.636', 'grad_norm': '52.82', 'learning_rate': '6.229e-06', 'epoch': '5'}
{'eval_loss': '2.507', 'eval_accuracy': '0.2248', 'eval_balanced_accuracy': '0.1975', 'eval_precision_macro': '0.148', 'eval_recall_macro': '0.1975', 'eval_f1_macro': '0.1414', 'eval_runtime': '1.283', 'eval_samples_per_second': '797.5', 'eval_steps_per_second': '24.95', 'epoch': '5'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.438', 'grad_norm': '36.81', 'learning_rate': '5.349e-06', 'epoch': '6'}
{'eval_loss': '2.485', 'eval_accuracy': '0.2229', 'eval_balanced_accuracy': '0.1937', 'eval_precision_macro': '0.1421', 'eval_recall_macro': '0.1937', 'eval_f1_macro': '0.1422', 'eval_runtime': '1.248', 'eval_samples_per_second': '819.5', 'eval_steps_per_second': '25.63', 'epoch': '6'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.277', 'grad_norm': '49.61', 'learning_rate': '4.462e-06', 'epoch': '7'}
{'eval_loss': '2.499', 'eval_accuracy': '0.2278', 'eval_balanced_accuracy': '0.1854', 'eval_precision_macro': '0.1304', 'eval_recall_macro': '0.1854', 'eval_f1_macro': '0.1423', 'eval_runtime': '1.244', 'eval_samples_per_second': '822.1', 'eval_steps_per_second': '25.72', 'epoch': '7'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'train_runtime': '214.9', 'train_samples_per_second': '266.6', 'train_steps_per_second': '8.375', 'train_loss': '4.877', 'epoch': '7'}


There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'eval_loss': '2.5', 'eval_accuracy': '0.2121', 'eval_balanced_accuracy': '0.1906', 'eval_precision_macro': '0.1561', 'eval_recall_macro': '0.1906', 'eval_f1_macro': '0.1497', 'eval_runtime': '1.509', 'eval_samples_per_second': '677.8', 'eval_steps_per_second': '21.2', 'epoch': '7'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.515', 'test_accuracy': '0.2219', 'test_balanced_accuracy': '0.1893', 'test_precision_macro': '0.1615', 'test_recall_macro': '0.1893', 'test_f1_macro': '0.1458', 'test_runtime': '1.314', 'test_samples_per_second': '778.6', 'test_steps_per_second': '24.36', 'epoch': '7'}


Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: roberta-base
Key                             | Status     | Details
--------------------------------+------------+--------
lm_head.layer_norm.bias         | UNEXPECTED |        
lm_head.bias                    | UNEXPECTED |        
lm_head.dense.bias              | UNEXPECTED |        
lm_head.dense.weight            | UNEXPECTED |        
roberta.embeddings.position_ids | UNEXPECTED |        
lm_head.layer_norm.weight       | UNEXPECTED |        
classifier.out_proj.weight      | MISSING    |        
classifier.dense.weight         | MISSING    |        
classifier.out_proj.bias        | MISSING    |        
classifier.dense.bias           | MISSING    |        

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
warmup_ratio is deprecated a


=== Training weightedCE_retry_lr2e-05 | lr=2e-05 | weighted=True ===
{'loss': '5.408', 'grad_norm': '5.152', 'learning_rate': '1.953e-05', 'epoch': '1'}
{'eval_loss': '2.708', 'eval_accuracy': '0.08211', 'eval_balanced_accuracy': '0.06667', 'eval_precision_macro': '0.005474', 'eval_recall_macro': '0.06667', 'eval_f1_macro': '0.01012', 'eval_runtime': '1.248', 'eval_samples_per_second': '819.5', 'eval_steps_per_second': '25.63', 'epoch': '1'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.351', 'grad_norm': '7.942', 'learning_rate': '1.775e-05', 'epoch': '2'}
{'eval_loss': '2.61', 'eval_accuracy': '0.2845', 'eval_balanced_accuracy': '0.1323', 'eval_precision_macro': '0.1119', 'eval_recall_macro': '0.1323', 'eval_f1_macro': '0.111', 'eval_runtime': '1.246', 'eval_samples_per_second': '820.9', 'eval_steps_per_second': '25.68', 'epoch': '2'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '5.039', 'grad_norm': '22.2', 'learning_rate': '1.598e-05', 'epoch': '3'}
{'eval_loss': '2.533', 'eval_accuracy': '0.1857', 'eval_balanced_accuracy': '0.1718', 'eval_precision_macro': '0.1368', 'eval_recall_macro': '0.1718', 'eval_f1_macro': '0.1242', 'eval_runtime': '1.264', 'eval_samples_per_second': '809.3', 'eval_steps_per_second': '25.32', 'epoch': '3'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.755', 'grad_norm': '24.94', 'learning_rate': '1.422e-05', 'epoch': '4'}
{'eval_loss': '2.516', 'eval_accuracy': '0.1965', 'eval_balanced_accuracy': '0.173', 'eval_precision_macro': '0.124', 'eval_recall_macro': '0.173', 'eval_f1_macro': '0.1312', 'eval_runtime': '1.263', 'eval_samples_per_second': '810', 'eval_steps_per_second': '25.34', 'epoch': '4'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.514', 'grad_norm': '41.5', 'learning_rate': '1.245e-05', 'epoch': '5'}
{'eval_loss': '2.494', 'eval_accuracy': '0.2131', 'eval_balanced_accuracy': '0.196', 'eval_precision_macro': '0.16', 'eval_recall_macro': '0.196', 'eval_f1_macro': '0.1432', 'eval_runtime': '1.257', 'eval_samples_per_second': '814', 'eval_steps_per_second': '25.46', 'epoch': '5'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '4.181', 'grad_norm': '37.08', 'learning_rate': '1.067e-05', 'epoch': '6'}
{'eval_loss': '2.545', 'eval_accuracy': '0.2287', 'eval_balanced_accuracy': '0.1756', 'eval_precision_macro': '0.1546', 'eval_recall_macro': '0.1756', 'eval_f1_macro': '0.1427', 'eval_runtime': '1.259', 'eval_samples_per_second': '812.6', 'eval_steps_per_second': '25.42', 'epoch': '6'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.884', 'grad_norm': '55.73', 'learning_rate': '8.913e-06', 'epoch': '7'}
{'eval_loss': '2.566', 'eval_accuracy': '0.2229', 'eval_balanced_accuracy': '0.2161', 'eval_precision_macro': '0.1711', 'eval_recall_macro': '0.2161', 'eval_f1_macro': '0.1531', 'eval_runtime': '1.257', 'eval_samples_per_second': '814', 'eval_steps_per_second': '25.46', 'epoch': '7'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.574', 'grad_norm': '34.02', 'learning_rate': '7.139e-06', 'epoch': '8'}
{'eval_loss': '2.617', 'eval_accuracy': '0.2248', 'eval_balanced_accuracy': '0.2074', 'eval_precision_macro': '0.1745', 'eval_recall_macro': '0.2074', 'eval_f1_macro': '0.1625', 'eval_runtime': '1.26', 'eval_samples_per_second': '811.7', 'eval_steps_per_second': '25.39', 'epoch': '8'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.306', 'grad_norm': '60.77', 'learning_rate': '5.366e-06', 'epoch': '9'}
{'eval_loss': '2.661', 'eval_accuracy': '0.2239', 'eval_balanced_accuracy': '0.2027', 'eval_precision_macro': '0.1735', 'eval_recall_macro': '0.2027', 'eval_f1_macro': '0.1644', 'eval_runtime': '1.25', 'eval_samples_per_second': '818.2', 'eval_steps_per_second': '25.59', 'epoch': '9'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '3.068', 'grad_norm': '44.93', 'learning_rate': '3.605e-06', 'epoch': '10'}
{'eval_loss': '2.677', 'eval_accuracy': '0.2375', 'eval_balanced_accuracy': '0.214', 'eval_precision_macro': '0.1885', 'eval_recall_macro': '0.214', 'eval_f1_macro': '0.1783', 'eval_runtime': '1.252', 'eval_samples_per_second': '816.9', 'eval_steps_per_second': '25.55', 'epoch': '10'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.892', 'grad_norm': '35.24', 'learning_rate': '1.832e-06', 'epoch': '11'}
{'eval_loss': '2.691', 'eval_accuracy': '0.2248', 'eval_balanced_accuracy': '0.2013', 'eval_precision_macro': '0.1782', 'eval_recall_macro': '0.2013', 'eval_f1_macro': '0.1694', 'eval_runtime': '1.267', 'eval_samples_per_second': '807.7', 'eval_steps_per_second': '25.26', 'epoch': '11'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'loss': '2.756', 'grad_norm': '30.43', 'learning_rate': '5.91e-08', 'epoch': '12'}
{'eval_loss': '2.713', 'eval_accuracy': '0.2209', 'eval_balanced_accuracy': '0.2051', 'eval_precision_macro': '0.1737', 'eval_recall_macro': '0.2051', 'eval_f1_macro': '0.1686', 'eval_runtime': '1.282', 'eval_samples_per_second': '797.9', 'eval_steps_per_second': '24.96', 'epoch': '12'}


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

{'train_runtime': '397.8', 'train_samples_per_second': '144', 'train_steps_per_second': '4.525', 'train_loss': '4.061', 'epoch': '12'}


There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

{'eval_loss': '2.677', 'eval_accuracy': '0.2375', 'eval_balanced_accuracy': '0.214', 'eval_precision_macro': '0.1879', 'eval_recall_macro': '0.214', 'eval_f1_macro': '0.1782', 'eval_runtime': '1.6', 'eval_samples_per_second': '639.3', 'eval_steps_per_second': '20', 'epoch': '12'}


early stopping required metric_for_best_model, but did not find eval_f1_macro so early stopping is disabled


{'test_loss': '2.731', 'test_accuracy': '0.2268', 'test_balanced_accuracy': '0.1968', 'test_precision_macro': '0.1901', 'test_recall_macro': '0.1968', 'test_f1_macro': '0.1669', 'test_runtime': '1.323', 'test_samples_per_second': '773.2', 'test_steps_per_second': '24.18', 'epoch': '12'}


,tag,lr,final_train_loss,val_f1_macro,val_balanced_acc,val_accuracy,test_f1_macro,test_balanced_acc,test_accuracy
1,weightedCE_retry_lr2e-05,0.00002,4.060733,0.178220,0.214001,0.237537,0.166871,0.196847,0.226784
0,weightedCE_retry_lr1e-05,0.00001,4.877433,0.149701,0.190620,0.212121,0.145798,0.189279,0.221896


## Step 5 — Combined results table and save

In [14]:
combined_rows = []
combined_rows.append({'model': 'TF-IDF (baseline, nb 02)', 'setup': '-', 'val_f1_macro': None, 'test_f1_macro': 0.134, 'test_bal_acc': None, 'test_acc': None})
combined_rows.append({'model': 'RoBERTa (04.1 original)', 'setup': 'weighted CE lr=2e-5 4ep', 'val_f1_macro': 0.0583, 'test_f1_macro': 0.0612, 'test_bal_acc': 0.0973, 'test_acc': 0.2776})

for r in sweep_results:
    combined_rows.append({'model': 'RoBERTa', 'setup': r['tag'], 'val_f1_macro': r['val_f1_macro'], 'test_f1_macro': r['test_f1_macro'], 'test_bal_acc': r['test_balanced_acc'], 'test_acc': r['test_accuracy']})

for r in weighted_retry_results:
    combined_rows.append({'model': 'RoBERTa', 'setup': r['tag'], 'val_f1_macro': r['val_f1_macro'], 'test_f1_macro': r['test_f1_macro'], 'test_bal_acc': r['test_balanced_acc'], 'test_acc': r['test_accuracy']})

combined_df = pd.DataFrame(combined_rows).sort_values('test_f1_macro', ascending=False, na_position='last').reset_index(drop=True)
print(combined_df.to_string())

with open(os.path.join(OUTPUT_DIR_ROOT, 'all_results.json'), 'w') as f:
    json.dump({
        'roberta_sweep': sweep_results,
        'roberta_weighted_retry': weighted_retry_results,
        'tfidf_baseline_f1_macro': 0.134,
    }, f, indent=2)
    
print('Saved all_results.json')

                      model                     setup  val_f1_macro  test_f1_macro  test_bal_acc  test_acc
0                   RoBERTa  weightedCE_retry_lr2e-05      0.178220       0.166871      0.196847  0.226784
1                   RoBERTa           plainCE_lr3e-05      0.181486       0.164361      0.174847  0.352884
2                   RoBERTa           plainCE_lr2e-05      0.164377       0.158272      0.171164  0.364614
3                   RoBERTa           plainCE_lr1e-05      0.144540       0.151584      0.170846  0.377322
4                   RoBERTa  weightedCE_retry_lr1e-05      0.149701       0.145798      0.189279  0.221896
5  TF-IDF (baseline, nb 02)                         -           NaN       0.134000           NaN       NaN
6   RoBERTa (04.1 original)   weighted CE lr=2e-5 4ep      0.058300       0.061200      0.097300  0.277600
7                   RoBERTa           plainCE_lr5e-05      0.030234       0.030156      0.066667  0.292278
Saved all_results.json
